### DISCUSSION TITLE
Mocking MqttClient

### DISCUSSION THREAD

**Discussion by ridomin** [dotnet/MQTTnet#1358] [General]
I was using `IMqttClient` from `v4-preview1` to create a  [MockMqttClient](https://github.com/ridomin/Rido.IoTClient/blob/master/Rido.IoTClient.Tests/MockMqttClient.cs) for unit tests.

With the latest `v4-preview5` the IMqttClient has been replaced by `sealed class MqttClient`, so I'm not able to create Mocks anymore, although in the README I still read: `Interfaces included for mocking and testing`

Can you consider removing the `sealed` limitation? Any other suggestion?

**Comment by chkr1011** [comment]
The interfaces are indeed gone in the next version so keep the API small. If you still want to replace the client etc. for mocking I recommend to add the interface on your own. Then you need to encapsulate the MQTT client in your custom wrapper. I will remove this information from the readme soon.

**Reply by mmuecke** [reply]
Why does the API become smaller by removing the interface? I do not understand the technical background and would appreciate an explanation. I am facing the same problem of not being able to create a mock anymore. What about removing the sealed? The same goes for the ManagedMqttClient.

If the implementation remains like this, I will provide a nuget that wraps the MqttClient and ManagedMqttClient classes via an interface.

**Reply by chkr1011** [reply]
The thing is that in the past we had to align the interface and the implementation including all overloads and comments. Sometimes methods were missing in the interface etc. After getting some inspiration from Azure SDK components like Service Bus SDK and the included clients I decided to get rid of the interfaces. HTTPClient, ServiceBusClient etc. also do not support such a thing. It is very uncommon to provide interfaces for such clients. 

Another "issue" is that the client and the managed client cannot share a common interface because the work completely different. When publishing with the regular client there is already a network transmit made and probably an exception thrown (that's why the method is called Publish). The managed client on the other hand only enqueues the packet internally. There will no exception at all. This is the wanted behavior but clearly a different behavior. 

The _sealed_ keyword is added to make sure that classes are not inherited in user code to keep breaking changes small when inheritance etc. is changed in the internals of the project. The sealed keyword can be for sure removed but my goal was to get rid of the maintenance effort.

Restoring a unique interface per client type is possible if the interest is big enough. The above explanation are just my thoughts.

**Reply by gfoidl** [reply]
> get rid of the interfaces. HTTPClient, ServiceBusClient etc. also do not support such a thing

`HttpClient` is a "bad example" here, as for testing it's very easy to have a test-`HttpMessageHandler` (or mock this one) and inject it into the `HttpClient`. So a different story, where an interface wouldn't bring any effort.

The possibility to test components is one point that speaks for interfaces.
Another point is the possibility to use dependency injection (together with testing, where a mock / stub gets injected).

> It is very uncommon to provide interfaces for such clients.

So this is IMO a suboptimal solution.

Without the interfaces, every user that has an eye for the aforementioned points, now needs to create a (custom) wrapper. This is effort that can be avoided.

I get your point about the maintenance cost. But shouldn't the public interface of a class become stable for a major version to avoid breaking changes?
And in editors like Visual Studio there's tooling support to extract an interface from a type. So maintenance cost isn't that much.

> The sealed keyword is added to make sure that classes are not inherited in user code to keep breaking changes small when inheritance etc.

`sealed` is a good point, also for performance as JIT nowadays and in the future can devirtualize more aggressively.
And you argument is very valid. I am all in for keeping the sealed. But only if interfaces are provided 😉.

**Reply by mmuecke** [reply]
Thanks @chkr1011  for the feedback and explanation. I support the suggestion of @gfoidl to introduce an interface for the two classes. For the documentation [inheritdoc](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#inheritdoc) can be used.

I can't contribute practical experience. But I tried a wrapper. A further problem is that all extensions on `MqttClient` (`MqttClientExtensions`) and `ManagedMqttClient` (`ManagedMqttClientExtensions`) are not available anymore. These are often used to publish something, for example. I find that already a big disadvantage by the lack of interfaces and wrapping the classes.

**Reply by chkr1011** [reply]
OK I will add the Interfaces again but keep the sealed Keywords. This seems to bei a good compromise to me. After that we can also discuss about the extensions.

**Reply by gfoidl** [reply]
Thanks @chkr1011!

**Reply by mmuecke** [reply]
Nice that the suggestion was accepted. I have created a pull request #1378

**Reply by chkr1011** [reply]
I added the interfaces again (see MyGet feed nuget). Please let me know if this works as expected for you.

**Reply by mmuecke** [reply]
Thank you. For me, testing works like this.

**Reply by mshwf** [reply]
I'm having the same issue with the `MqttServer`, why its interface wasn't restore like the `MqttClient`?
